### PR TITLE
implement one key from keylist is needed in order to delete file, and delete E2E tests

### LINKED ISSUE #8054: FileDeleteSuite

### PR DESCRIPTION
**Pull request by povolev15**
implement one key from key list is needed in order to delete file using the threshold key, and  make the delete-suite working properly

**Related issue(s)**:

Fixes #8054

### CODE REVIEW

File: hedera-node/test-clients/src/eet/java/EndToEndTests.java [hunk 1]
```diff
@@ -65,9 +69,9 @@ Collection<DynamicContainer> file() {
                 //				extractSpecsFromSuite(DiverseStateValidation::new),
                 extractSpecsFromSuite(ExchangeRateControlSuite::new),
                 //              extractSpecsFromSuite(FetchSystemFiles::new),
-                //              extractSpecsFromSuite(FileAppendSuite::new),
-                //              extractSpecsFromSuite(FileCreateSuite::new),
-                //				extractSpecsFromSuite(FileDeleteSuite::new),
+                extractSpecsFromSuite(FileAppendSuite::new),
+                extractSpecsFromSuite(FileCreateSuite::new),
+                extractSpecsFromSuite(FileDeleteSuite::new),
                 extractSpecsFromSuite(FileUpdateSuite::new),
```
Comment (jsync-swirlds) [inline hedera-node/test-clients/src/eet/java/EndToEndTests.java:75]: Good to see some of these re-enabled.
Always been a little odd that so many E2E tests are commented out.
Perhaps @rbair23 or @tinker-michaelj has some insight to why so few of these tests are enabled.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java [hunk 2]
```diff
@@ -158,7 +158,7 @@ void preHandleWorksAsExpected() throws PreCheckException {
         subject.preHandle(realPreContext);
 
         assertTrue(realPreContext.requiredNonPayerKeys().size() > 0);
-        assertEquals(3, realPreContext.requiredNonPayerKeys().size());
+        assertEquals(1, realPreContext.requiredNonPayerKeys().size());
```
Comment (jsync-swirlds) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java:161]: Might want to add assertions here that the 1 required key is a `Threshold` key with threshold `1`.

**Review by jsync-swirlds — APPROVED** [review]

**Review by vtronkov — APPROVED** [review]

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java [hunk 3]
```diff
@@ -50,27 +49,21 @@ public static void main(String... args) {
...
+        return List.of(getDeletedFileInfo(), canDeleteWithAnyOneOfTopLevelKeyList());
     }
 
+    @HapiTest
     private HapiSpec canDeleteWithAnyOneOfTopLevelKeyList() {
         KeyShape shape = listOf(SIMPLE, threshOf(1, 2), listOf(2));
         SigControl deleteSigs = shape.signedWith(sigs(ON, sigs(OFF, OFF), sigs(ON, OFF)));
 
-        return defaultFailingHapiSpec("CanDeleteWithAnyOneOfTopLevelKeyList")
+        return defaultHapiSpec("CanDeleteWithAnyOneOfTopLevelKeyList")
```
Comment (iwsimon) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: Why is the name changed here? Did you run the mono code test to make sure this is correct with this change?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 4]
```diff
@@ -124,6 +125,36 @@ public static void validateAndAddRequiredKeys(
...
+            KeyList fileKeyList = file.keys();
+
+            if (fileKeyList != null && fileKeyList.hasKeys()) {
+                ThresholdKey syntheticKey =
+                        ThresholdKey.newBuilder().threshold(1).keys(fileKeyList).build();
+                context.requireKey(Key.newBuilder().thresholdKey(syntheticKey).build());
+            }
+        }
+
+        if (transactionKeys != null && transactionKeys.hasKeys()) {
```
Comment (iwsimon) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:151]: What's transaction keys?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java (hunk 3 above)
Comment (povolev15) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: because it was part of the negative group and we moved it to be part of the positive otherwise it passed but Intelliji show it as failed.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 4 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:151]: These are the keys in context.body().keys() and need to be added as well.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java (hunk 2 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java:161]: sure will be done

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 4 above)
Comment (iwsimon) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:151]: context.body(): TransactionBody, I don't see any keys in TransactionBody.

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:151]: you are right for delete not exist we send null I will remove and fix it right away. thank you great catch

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java (hunk 3 above)
Comment (iwsimon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: Did you try to run, application.run (old mono code way to test), then run FileDeletSuite.main() to see if the test still pass?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java [hunk 5]
```diff
@@ -68,7 +68,7 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         preValidate(transactionBody.fileID(), fileStore, context, true);
 
         var file = fileStore.getFileLeaf(transactionBody.fileID());
-        validateAndAddRequiredKeys(file.orElse(null), null, context);
+        validateAndAddRequiredKeysForDelete(file.orElse(null), null, context);
```
Comment (netopyr) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java]: Not part of this PR, but why does `ReadableFileStore.getFileLeaf()` return an `Optional`? We had a lengthy discussion at the start of the project and decided to return `null` in such cases. This code would be a lot simpler.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java [hunk 6]
```diff
@@ -158,7 +158,14 @@ void preHandleWorksAsExpected() throws PreCheckException {
...
```
Comment (netopyr) [inline hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java:161]: Please use `AssertJ` for assertions

File: hedera-node/test-clients/src/eet/java/EndToEndTests.java (hunk 1 above)
Comment (netopyr) [reply hedera-node/test-clients/src/eet/java/EndToEndTests.java:75]: Do these tests also succeed in all other environments where these tests are used?

**Review by vtronkov — APPROVED** [review]

**Review by netopyr — CHANGES_REQUESTED** [review]
It worries me that the tests and how they are executed is changed here. This is certainly our goal, but we must ensure that the mono tests are successful, too.

I suggest to split the actual fix from all changes regarding the test specification. We need to talk to @tinker-michaelj and @rbair23 before enabling them.

**Comment by povolev15** [comment]
> It worries me that the tests and how they are executed is changed here. This is certainly our goal, but we must ensure that the mono tests are successful, too.
> 
> I suggest to split the actual fix from all changes regarding the test specification. We need to talk to @tinker-michaelj and @rbair23 before enabling them.

@netopyr it is passing in CI. As I understand, it runs the mono code for these tests.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java]: ok I will fix it no worries. maybe I missed it same happens on consensus service as well.
I can also fix it here it is not an issue at all.

File: hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java (hunk 6 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/test/java/com/hedera/node/app/service/file/impl/test/handlers/FileDeleteTest.java:161]: sure I will use assertj

**Comment by povolev15** [comment]
>Do these tests also succeed in all other environments where these tests are used?
@netopyr  I'm not sure if I understand this question it passes the CI, it is not annotated as @HapiSuites.
I thought we were supposed to each suite that passed to enable in `EndToEnd` class.

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java]: created a separate issue for it https://github.com/hashgraph/hedera-services/issues/8198

File: hedera-node/test-clients/src/eet/java/EndToEndTests.java (hunk 1 above)
Comment (povolev15) [reply hedera-node/test-clients/src/eet/java/EndToEndTests.java]: I removed it, and we will address it in separate PR after a discussion with @tinker-michaelj

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java (hunk 3 above)
Comment (iwsimon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: I thought we were not supposed to change the test code itself besides adding @HapiTest after each test was passed. Yes, this one is using defaultFailingHapiSpec() instead of defaultHapiSpec(). There are other test suite also use this. We probably should add this handling in HapiTestEngine instead of changing the suite test code.

Comment (jsync-swirlds) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: It's not defaultFailingHapiSpec that's the problem here.  It's the combination of using defaultFailingHapiSpec and also placing the test in `negativeTests`.  That combination creates a situation where the test reports FAILURE, but succeeds, and reports SUCCESS, but fails.

We cannot make HapiTestEngine do that without some really ugly changes, and it's a bad setup regardless as it always reports the opposite of what is true (which creates a high probability someone eventually "fixes" it by making the code do the opposite of intended).

Yes, we should not change tests, in general, but this is an exception case where the test was setup badly to begin with, and all we do here is fix that setup to a sane structure so that both HapiTest and MonoService either succeed together or fail together.

Comment (iwsimon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: negativeTests() doesn't do anything except putting the test in the Arrays and return a list, and probably telling you the test expects "expected.final.status" will be "FAILED" as defaultFailingHapiSpec() uses. Default "expected.final.status" is "PASSED" in which defaultHapiSpec() uses. When you run this in mono code, you don't see FAILURE, but succeeds and only see this in HapiTestEngine run. If you change code this way, there are other two suite tests will have to do the same.

Comment (jsync-swirlds) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: We observed different behavior, the test was consistently reporting failure, regardless of how it was run, but the suite succeeded.  The cleaned up lists above fixed that behavior so it failed if it reported FAILURE, and succeeded if it reported SUCCESS.  Then this line changed to make it actually succeed when it was supposed to succeed and fail when it was supposed to fail.
Not sure what dark ritual in the HapiSpecSuite was causing all that, but this set of changes fixed it.
The other suites only have to change if they are failing incorrectly.  This change was necessary, not simply desired.

Comment (jsync-swirlds) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: Here is the prior behavior:
```
2023-08-24 12:42:51.145 INFO   282  HapiTxnOp - 'CanDeleteWithAnyOneOfTopLevelKeyList' -  HapiFileDelete{sigs=3, node=0.0.3, fileName=test} Wrong actual status INVALID_SIGNATURE, expected SUCCESS
2023-08-24 12:42:51.145 WARN   262  HapiSpecOperation - 'CanDeleteWithAnyOneOfTopLevelKeyList' - HapiFileDelete{sigs=3, node=0.0.3, fileName=test} failed - Wrong status! Expected SUCCESS, was INVALID_SIGNATURE!
2023-08-24 12:42:51.482 INFO   481  HapiSpec - 'CanDeleteWithAnyOneOfTopLevelKeyList' - final status: FAILED!
```
Now perhaps this was just broken in several levels (and it might not even run in current E2E or integration tests).  Regardless the change Lev made here seems to make the test correct and run correctly in both Modularized and Mono Services.

Comment (iwsimon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: This is the output from HapiTestEngine run.

Comment (jsync-swirlds) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: The PASSED_UNEXPECTEDLY is what was causing issues, yes.
We couldn't figure out what made that translation from "this should have failed, but did not fail" to "it's fine".
Interesting, the Mono code returns a SUCCESS value if you debug it, rather than INVALID_SIGNATURE, so something was causing `defaultFailingHapiSpec` to accept a success on the test that *should have failed* and translating to PASSED_UNEXPECTEDLY, which is kind of "we should have failed, but didn't, so we'll say the test passed" and seems completely broken.
If we matched the modular code to mono, the HapiTest failed, if we returned the wrong thing, the HapiTest succeeded, and that was clearly not correct.
Changing the spec call to `defaultHapiSpec` makes BOTH versions return PASSED (neither returns PASSED_UNEXPECTEDLY), and that seems more correct than one passing through a backdoor that causes the other to accept incorrect code.

If we have other PASSED_UNEXPECTEDLY items, we definitely should investigate to see if something is broken in the test definition (as was the case here) or if we have a possible bug that the testing doesn't report (because PASSED_UNEXPECTEDLY is not reported as a failure).

Comment (jsync-swirlds) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: Short summary:  This test was unequivocally broken and regardless of HapiTest or not, needed to be fixed.  It is fixed now.
While we do not want to change the test suites unless absolutely necessary, it is necessary in this case.

Comment (iwsimon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: I think we should check with @tinker-michaelj or @Neeharika-Sompalli about this defaultFailingHapiSpec().

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java (hunk 5 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/handlers/FileDeleteHandler.java]: already merged

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java [hunk 7]
```diff
@@ -124,6 +125,28 @@ public static void validateAndAddRequiredKeys(
...
+            KeyList fileKeyList = file.keys();
+
+            if (fileKeyList != null && fileKeyList.hasKeys()) {
+                // this threshold created to solve the issue of the file delete that can be deleted using one of the
+                // keys in the keylist that's why it is wrapped in threshold key with threshold 1
+                ThresholdKey syntheticKey =
+                        ThresholdKey.newBuilder().threshold(1).keys(fileKeyList).build();
+                context.requireKey(Key.newBuilder().thresholdKey(syntheticKey).build());
+            }
+        }
```
Comment (rbair23) [inline hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: I think there is a bug here. If File is null, or if it has no keys, then we don't add a "requiredKey". But then this means it is by default accepted. So if the File is null for some reason, the transaction will succeed? Or if the file has no keys on it, instead of forbidding modification, it will be allowed. Is that right?

Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: in File handle we check `validateFalse(file.keys() == null, UNAUTHORIZED);`
The function you commented about is just for prehandle when we add required keys.

Comment (tinker-michaelj) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: It seems like we could just throw a `PreCheckException(INVALID_FILE_ID)` on a missing `File`, as e.g. `TokenDeleteHandler` does [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenDeleteHandler.java#L60).

**Review by tinker-michaelj — DISMISSED** [review]
I think it's slightly clearer to throw `PreCheckException` on a missing target file, but otherwise LGTM!

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java (hunk 3 above)
Comment (povolev15) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: @tinker-michaelj can you comment on it?

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 7 above)
Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: This isn't checking for missing file, it's handling that delete only requires a single key to succeed.  Missing file is checked elsewhere, and already throws as described.

**Review by jsync-swirlds — APPROVED** [review]

File: hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java (hunk 7 above)
Comment (povolev15) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: That exactly what is happens in the function that was called before that
`preValidate(transactionFileId, fileStore, context, true);`

``` public static void preValidate(
            @Nullable final FileID fileId,
            @NonNull final ReadableFileStore fileStore,
            @NonNull final PreHandleContext context,
            boolean isDelete)
            throws PreCheckException {
        requireNonNull(context);

        if (fileId == null) {
            throw new PreCheckException(INVALID_FILE_ID);
        }

        final var fileConfig = context.configuration().getConfigData(FilesConfig.class);```

Comment (netopyr) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: @jsync-swirlds 
Regarding your concerns, the list of required keys we get in pre-handle only triggers the signature verification asynchronously. We do not use this information for anything else. In handle, we get the required keys again (this time from the right state), and based on that information, we verify all required keys are valid.

Comment (jsync-swirlds) [reply hedera-node/hedera-file-service-impl/src/main/java/com/hedera/node/app/service/file/impl/utils/FileServiceUtils.java:147]: Ahh, that's the second call to preHandle in the Handle flow.  I forgot about that.
Thank you, @netopyr .

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: @povolev15 the new test looks correct to me. 💯

**Review by netopyr — APPROVED** [review]
LGTM

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java (hunk 3 above)
Comment (povolev15) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: @iwsimon please let me know if I can resolve it

Comment (iwsimon) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/file/FileDeleteSuite.java:60]: LGTM

**Review by iwsimon — APPROVED** [review]
